presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Config.java

```diff
@@ -65,6 +65,8 @@
     private PrestoS3AclType s3AclType = PrestoS3AclType.PRIVATE;
     private boolean skipGlacierObjects;
     private boolean requesterPaysEnabled;
+    private boolean s3StreamingUploadEnabled;
```

Comment:
Should we enable this by default? Are you using this change in production already?

Reply:
I prefer not enable this feature by default.

We are using this feature in production already, but we are using it without large loading.

And there is concern that user should create a [AbortIncompleteMultipartUpload](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-abortincompletemultipartupload.html) policy, so they don't get charged if any upload is neither uploaded successfully nor never aborted (might caused by server crashed during upload).
Our company create the policy automatically when the S3FileSystem was initialized.

Reply:
We already use multipart uploads via `TransferManager`, so the policy would be needed in either case. It's unfortunate that they don't have a default policy that aborts after some number of days. It seems ridiculous that the default is to keep them around forever.

I'm ok with not enabling by default for now. We can do that in a future release, after we have confirmation that it is working well for multiple users.

One thing I just realized is that we have `hive.s3.multipart.min-file-size` for the existing behavior. We could add similar but simpler logic for the streaming approach to not use multipart if there would only be one part (i.e. if we are flushing when finished and there are no parts yet). This can be a follow up.

Reply:
I created a PR to implement this: https://github.com/prestosql/presto/pull/6201

It also improves the abort logic. I realized your original implementation of aborting immediately was good:

* ensure we cleanup even if the output stream is not closed for some reason (bugs elsewhere)
* since `close()` is called after a failure, we need to ensure that we don't finish the upload

I added a `failed` flag to track the fact that a previous part upload failed, since the abort might fail but the finish could succeed.